pkg/pillar/cmd/zedagent/localinfo.go

```diff
@@ -192,73 +195,133 @@ func processReceivedAppCommands(ctx *getconfigContext, cmdList *profile.LocalApp
 			continue
 		}
 		// Try to find the application instance.
-		ais := findAppInstance(ctx, appUUID, displayName)
-		if ais == nil {
+		appInst := findAppInstance(ctx, appUUID, displayName)
+		if appInst == nil {
 			log.Warnf("Failed to find app instance with UUID=%s, displayName=%s",
 				appUUID, displayName)
 			continue
 		}
-		appUUID = ais.UUIDandVersion.UUID
+		appUUID = appInst.UUIDandVersion.UUID
+		if _, duplicate := publishApps[appUUID.String()]; duplicate {
+			log.Warnf("Multiple commands requested for app instance with UUID=%s",
+				appUUID)
+			continue
+		}
+
 		command := types.AppCommand(appCmdReq.Command)
-		appCmd := ctx.localAppCommands.LookupByAppUUID(appUUID)
-		if appCmd != nil {
-			// Entry for this app already exists.
-			if appCmd.Command == command &&
-				appCmd.LocalServerTimestamp == appCmdReq.Timestamp {
-				// already accepted
-				continue
+		appCmd := ctx.localAppCommands[appUUID.String()]
+		if appCmd == nil {
+			appCmd = &types.LocalAppCommand{
+				AppUUID:          appUUID,
+				VolumeGenCounter: make(map[string]int64),
 			}
-			appCmd.Command = command
-			appCmd.LocalServerTimestamp = appCmdReq.Timestamp
-			appCmd.DeviceTimestamp = time.Now()
-			appCmd.Completed = false
-			cmdChanges = true
+			ctx.localAppCommands[appUUID.String()] = appCmd
+		}
+		if appCmd.Command == command &&
+			appCmd.LocalServerTimestamp == appCmdReq.Timestamp {
+			// already accepted
 			continue
 		}
-		// Add new entry.
-		ctx.localAppCommands.Cmds = append(ctx.localAppCommands.Cmds,
-			types.LocalAppCommand{
-				AppUUID:              appUUID,
-				Command:              command,
-				LocalServerTimestamp: appCmdReq.Timestamp,
-				DeviceTimestamp:      time.Now(),
-				Completed:            false,
-			})
+		appCmd.Command = command
+		appCmd.LocalServerTimestamp = appCmdReq.Timestamp
+		appCmd.DeviceTimestamp = time.Now()
+		appCmd.Completed = false
+
+		// Update configuration to trigger the operation.
+		switch appCmd.Command {
+		case types.AppCommandRestart:
+			appCmd.RestartCounter++
+			appInst.RestartCmd.Counter++
+			publishApps[appUUID.String()] = *appInst
+
+		case types.AppCommandPurge:
+			appCmd.PurgeCounter++
+			appInst.PurgeCmd.Counter++
```

Comment:
When we get a restart or purge command (well, actally an update to the AppInstanceConfig) from the controller we will check against the previous counters from the controller. Won't this local restart/purge command upset that check since they appear to increment the counters (as opposed to a separate Local*Counter?)

Reply:
Yeah this is very confusing.

IMHO zedmanager should operate on the AppInstanceConfig received from the controller.
Thus zedmanager would need to add in the local cmd counters as appropriate.

The original code (where zedagent looks at AppInstanceStatus to report back to the LPS) is fine, but purging/restarting is sufficiently complex to not have that logic spread between zedagent and zedmanager.